In [30]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)

In [14]:
df = pd.read_csv("../data/WA_Fn-UseC_-HR-Employee-Attrition.csv")

In [15]:
import os
os.getcwd()

'D:\\1\\HR-Analytics-Employee-Attrition\\notebooks'

In [16]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,Sales Executive,4,Single,5993,19479,8,Y,Yes,11,3,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,Research Scientist,2,Married,5130,24907,1,Y,No,23,4,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,Laboratory Technician,3,Single,2090,2396,6,Y,Yes,15,3,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,Research Scientist,3,Married,2909,23159,1,Y,Yes,11,3,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,Laboratory Technician,2,Married,3468,16632,9,Y,No,12,3,4,80,1,6,3,3,2,2,2,2


In [17]:
df.tail()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1465,36,No,Travel_Frequently,884,Research & Development,23,2,Medical,1,2061,3,Male,41,4,2,Laboratory Technician,4,Married,2571,12290,4,Y,No,17,3,3,80,1,17,3,3,5,2,0,3
1466,39,No,Travel_Rarely,613,Research & Development,6,1,Medical,1,2062,4,Male,42,2,3,Healthcare Representative,1,Married,9991,21457,4,Y,No,15,3,1,80,1,9,5,3,7,7,1,7
1467,27,No,Travel_Rarely,155,Research & Development,4,3,Life Sciences,1,2064,2,Male,87,4,2,Manufacturing Director,2,Married,6142,5174,1,Y,Yes,20,4,2,80,1,6,0,3,6,2,0,3
1468,49,No,Travel_Frequently,1023,Sales,2,3,Medical,1,2065,4,Male,63,2,2,Sales Executive,2,Married,5390,13243,2,Y,No,14,3,4,80,0,17,3,2,9,6,0,8
1469,34,No,Travel_Rarely,628,Research & Development,8,3,Medical,1,2068,2,Male,82,4,2,Laboratory Technician,3,Married,4404,10228,2,Y,No,12,3,1,80,0,6,3,4,4,3,1,2


In [18]:
df.sample(5,random_state=42)

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1041,28,No,Travel_Rarely,866,Sales,5,3,Medical,1,1469,4,Male,84,3,2,Sales Executive,1,Single,8463,23490,0,Y,No,18,3,4,80,0,6,4,3,5,4,1,3
184,53,No,Travel_Rarely,1084,Research & Development,13,2,Medical,1,250,4,Female,57,4,2,Manufacturing Director,1,Divorced,4450,26250,1,Y,No,11,3,3,80,2,5,3,3,4,2,1,3
1222,24,Yes,Travel_Rarely,240,Human Resources,22,1,Human Resources,1,1714,4,Male,58,1,1,Human Resources,3,Married,1555,11585,1,Y,No,11,3,3,80,1,1,2,3,1,0,0,0
67,45,No,Travel_Rarely,1339,Research & Development,7,3,Life Sciences,1,86,2,Male,59,3,3,Research Scientist,1,Divorced,9724,18787,2,Y,No,17,3,3,80,1,25,2,3,1,0,0,0
220,36,No,Travel_Rarely,1396,Research & Development,5,2,Life Sciences,1,304,4,Male,62,3,2,Laboratory Technician,2,Single,5914,9945,8,Y,No,16,3,4,80,0,16,3,4,13,11,3,7


## Observation

The dataset loads correctly. The sample confirms that employee records span multiple departments, job roles, and demographic characteristics. Column names appear consistent and ready for further inspection.

In [20]:
df.shape

(1470, 35)

### Observation

The dataset contains information for 1,470 employees with 35 variables. This provides sufficient data for workforce and attrition analysis.

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

### Observation

The dataset contains both numerical and categorical variables. All columns have complete records with no immediate indication of missing values. The data types appear appropriate for the stored values.

In [25]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.923810,9.135373,18.0,30.00,36.0,43.00,60.0
DailyRate,1470.0,802.485714,403.509100,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.192517,8.106864,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.912925,1.024165,1.0,2.00,3.0,4.00,5.0
EmployeeCount,1470.0,1.000000,0.000000,1.0,1.00,1.0,1.00,1.0
EmployeeNumber,1470.0,1024.865306,602.024335,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.721769,1.093082,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.891156,20.329428,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.729932,0.711561,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.063946,1.106940,1.0,1.00,2.0,3.00,5.0


In [24]:
df.describe(include="object").T

,count,unique,top,freq
Attrition,1470,2,No,1233
BusinessTravel,1470,3,Travel_Rarely,1043
Department,1470,3,Research & Development,961
EducationField,1470,6,Life Sciences,606
Gender,1470,2,Male,882
JobRole,1470,9,Sales Executive,326
MaritalStatus,1470,3,Married,673
Over18,1470,1,Y,1470
OverTime,1470,2,No,1054


## Observation

The dataset contains a mix of numerical and categorical employee information. The average employee age is approximately 37 years, with salaries ranging from 1,009 to 19,999 per month. Research & Development is the largest department, while Sales Executive is the most common job role. Most employees have not left the organization, indicating that the dataset is imbalanced toward retained employees. These characteristics provide a strong foundation for analyzing factors that influence employee attrition.

In [26]:
# Missing values
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [27]:
# Duplicate records
df.duplicated().sum()

np.int64(0)

In [28]:
df.nunique()

Age                           43
Attrition                      2
BusinessTravel                 3
DailyRate                    886
Department                     3
DistanceFromHome              29
Education                      5
EducationField                 6
EmployeeCount                  1
EmployeeNumber              1470
EnvironmentSatisfaction        4
Gender                         2
HourlyRate                    71
JobInvolvement                 4
JobLevel                       5
JobRole                        9
JobSatisfaction                4
MaritalStatus                  3
MonthlyIncome               1349
MonthlyRate                 1427
NumCompaniesWorked            10
Over18                         1
OverTime                       2
PercentSalaryHike             15
PerformanceRating              2
RelationshipSatisfaction       4
StandardHours                  1
StockOptionLevel               4
TotalWorkingYears             40
TrainingTimesLastYear          7
WorkLifeBa

## Observation

The dataset contains several columns with only a single unique value (`EmployeeCount`, `StandardHours`, and `Over18`), indicating that they do not provide meaningful variation for analysis. Other categorical variables, such as `Department`, `Gender`, and `JobRole`, contain a reasonable number of distinct categories and are expected to be valuable for exploratory analysis.

# Data Dictionary

| Column | Type | Business Meaning | Keep? |
|---------|------|------------------|--------|
| Age | Numerical | Employee's age in years | ✅ |
| Attrition | Target | Whether the employee left the organization | ✅ |
| BusinessTravel | Categorical | Frequency of work-related travel | ✅ |
| DailyRate | Numerical | Employee's daily pay rate | ✅ |
| Department | Categorical | Department where the employee works | ✅ |
| DistanceFromHome | Numerical | Distance between home and workplace | ✅ |
| Education | Ordinal | Education level (1–5) | ✅ |
| EducationField | Categorical | Field of education | ✅ |
| EmployeeCount | Constant | Always equals 1 | ❌ Remove |
| EmployeeNumber | Identifier | Unique employee ID | ❌ Remove |
| EnvironmentSatisfaction | Ordinal | Satisfaction with work environment (1–4) | ✅ |
| Gender | Categorical | Employee gender | ✅ |
| HourlyRate | Numerical | Hourly pay rate | ✅ |
| JobInvolvement | Ordinal | Level of job involvement (1–4) | ✅ |
| JobLevel | Ordinal | Employee seniority level (1–5) | ✅ |
| JobRole | Categorical | Employee's job designation | ✅ |
| JobSatisfaction | Ordinal | Satisfaction with the job (1–4) | ✅ |
| MaritalStatus | Categorical | Marital status | ✅ |
| MonthlyIncome | Numerical | Monthly salary | ✅ |
| MonthlyRate | Numerical | Monthly pay rate | ✅ |
| NumCompaniesWorked | Numerical | Number of companies worked at previously | ✅ |
| Over18 | Constant | Always "Y" | ❌ Remove |
| OverTime | Categorical | Whether employee works overtime | ✅ |
| PercentSalaryHike | Numerical | Percentage salary increase | ✅ |
| PerformanceRating | Ordinal | Performance rating (3–4) | ✅ |
| RelationshipSatisfaction | Ordinal | Satisfaction with workplace relationships | ✅ |
| StandardHours | Constant | Always 80 hours | ❌ Remove |
| StockOptionLevel | Ordinal | Stock option benefit level | ✅ |
| TotalWorkingYears | Numerical | Total years of work experience | ✅ |
| TrainingTimesLastYear | Numerical | Number of training sessions last year | ✅ |
| WorkLifeBalance | Ordinal | Work-life balance rating (1–4) | ✅ |
| YearsAtCompany | Numerical | Years spent in the current company | ✅ |
| YearsInCurrentRole | Numerical | Years in the current role | ✅ |
| YearsSinceLastPromotion | Numerical | Years since last promotion | ✅ |
| YearsWithCurrManager | Numerical | Years working with the current manager | ✅ |

## Data Summary 
Some columns, such as `EmployeeCount`, `Over18`, and `StandardHours`, contain only a single value for all records, while `EmployeeNumber` functions as a unique identifier. These columns do not contribute meaningful information to the analysis and will be removed during the data cleaning phase.

## Business Objective

The objective of this project is to identify the key factors that contribute to employee attrition. The analysis focuses on employee demographics, job roles, salary, work environment, overtime, and satisfaction levels to understand patterns associated with employee turnover.

The insights generated from this analysis can support data-driven HR decisions and help improve employee retention strategies.